File: frontend/src/api/settingAgent.ts
/**
 * Setting Agent API
 * 设定管理 Agent 接口
 */

import { api } from './client'
import type { AssistantContextSummary } from './assistantContext'

const API_BASE = '/setting-agent'

// ==================== 类型定义 ====================

export interface CharacterReferenceCandidate {
  character_id: string
  name: string
  role?: string
  importance_tier?: string
  description?: string
  match_reason: string
  confidence: number
}

export interface ResolvedCharacterReference {
  status: 'resolved'
  source_text: string
  source_payload?: unknown
  character_id: string
  character_name: string
  confidence: number
  resolution_method: string
  provenance?: Record<string, unknown>
}

export interface UnresolvedCharacterReference {
  status: 'unresolved' | 'ambiguous'
  source_text: string
  source_payload?: unknown
  reason: string
  message: string
  candidates: CharacterReferenceCandidate[]
  recommended_actions: Array<'bind_existing' | 'create_character' | 'keep_text_only' | 'ignore'>
  provenance?: Record<string, unknown>
}

export interface CharacterReferenceResolution {
  success?: boolean
  related_characters: string[]
  related_character_refs: ResolvedCharacterReference[]
  unresolved_character_refs: UnresolvedCharacterReference[]
  has_unresolved: boolean
}

export interface PendingLore {
  title: string
  category: string
  priority: string
  content: string
  summary: string
  keywords: string[]
  tags: string[]
  constraints: string[]
  related_characters: string[]
  related_character_refs?: ResolvedCharacterReference[]
  unresolved_character_refs?: UnresolvedCharacterReference[]
  related_locations: string[]
  related_items: string[]
  related_factions?: string[]
  depends_on_lore?: string[]
  supports_lore?: string[]
  potential_conflicts?: string[]
  usage_guidance?: string
  resource_requirements?: Array<Record<string, unknown>>
}

export interface PendingCharacter {
  id?: string
  name: string
  importance_tier: string
  description: string
  appearance: string
  personality: string
  background_story: string
  speech_pattern: string
  age: number | null
  gender: string
  goals: string[]
  relationships?: string[]
  key_relationships?: Record<string, string>
  lexicon?: string[]
  forbidden_words?: string[]
  voice_samples?: string[]
  attributes?: Record<string, unknown>
  inventory?: string[]
  narrative_weight?: string
  story_arc_role?: string
  plot_priority?: number
  has_agent?: boolean
  agent_enabled?: boolean
  agent_goals?: string[]
  agent_memory?: string[]
}

export interface PendingHook {
  title: string
  description: string
  hook_type: 'mystery' | 'object' | 'character' | 'event' | 'location' | 'relationship' | 'custom'
  status: 'planted' | 'triggered' | 'resolved' | 'dropped'
  related_characters: string[]
  related_locations: string[]
  related_objects: string[]
  plant_context: string
  resolution_hint: string
  priority: number
}

export interface ChatResponse {
  message: string
  session_id: string
  mode: string
  structured_data?: Record<string, unknown>
  lore_saved?: boolean
  pending_lores?: PendingLore[]
  pending_characters?: PendingCharacter[]
  pending_hooks?: PendingHook[]
  cached_pending_lores?: PendingLore[]
  cached_pending_characters?: PendingCharacter[]
  cached_pending_hooks?: PendingHook[]
  conversation_history?: Array<{
    role: 'user' | 'assistant'
    content: string
    timestamp?: string
  }>
  improvement_suggestions?: ImprovementSuggestion[]
  assistant_session_id?: string
  context_packet?: AssistantContextSummary
}

export interface ImprovementSuggestion {
  id: string
  type: 'conflict' | 'missing' | 'priority' | 'optimize' | 'relation'
  target_lore_id?: string
  target_lore_title: string
  issue: string
  suggestion: string
  suggested_content?: string
  priority: 'low' | 'medium' | 'high'
  reason: string
}

export interface SettingConflict {
  id: string
  conflict_type: string
  description: string
  severity: 'low' | 'medium' | 'high' | 'critical'
  existing_lore_id?: string
  existing_lore_title?: string
  new_lore_title?: string
  status: 'pending' | 'negotiating' | 'resolved' | 'escalated'
  resolution_suggestions: string[]
  selected_resolution?: string
  detected_at: string
  resolved_at?: string
}

export interface SettingChangeResult {
  success: boolean
  request: {
    id: string
    project_id: string
    change_type: string
    target_lore_id?: string
    new_lore_data?: Record<string, unknown>
    conflicts: SettingConflict[]
    has_conflicts: boolean
    status: string
    created_at: string
  }
  has_conflicts: boolean
  conflicts: SettingConflict[]
}

export interface NegotiateResult {
  status: 'resolved' | 'negotiating'
  conflict?: SettingConflict
  can_proceed: boolean
  message?: string
  suggestions?: string[]
}

export interface ExecuteResult {
  success: boolean
  result?: Record<string, unknown>
  error?: string
  unresolved_conflicts?: SettingConflict[]
}

export interface LoreSummary {
  project_id: string
  total_lore_count: number
  by_category: Record<string, number>
  by_priority: Record<string, number>
  constitutional_rules: string[]
  key_entities: string[]
  recent_changes: Array<Record<string, unknown>>
  pending_conflicts: number
  generated_at: string
}

export interface ConversationHistory {
  project_id?: string
  session_id: string
  messages?: Array<{
    role: 'user' | 'assistant'
    content: string
    timestamp?: string
    created_at?: string
  }>
  history?: Array<{
    role: 'user' | 'assistant'
    content: string
    timestamp: string
  }>
  pending_lores?: PendingLore[]
  pending_characters?: PendingCharacter[]
  pending_hooks?: PendingHook[]
  context_packet?: AssistantContextSummary | null
  assistant_session_id?: string
  total?: number
}

export interface PendingConflicts {
  project_id: string
  pending_conflicts: SettingConflict[]
  count: number
}

// ==================== API 函数 ====================

/**
 * 与 Setting Agent 聊天
 */
export async function chatWithSettingAgent(
  projectId: string,
  message: string,
  context?: Record<string, unknown>,
  options?: { sessionId?: string; requestId?: string },
): Promise<ChatResponse> {
  return await api.post(`${API_BASE}/chat`, {
    project_id: projectId,
    message,
    context,
    session_id: options?.sessionId,
    request_id: options?.requestId,
  })
}

/**
 * 请求设定变更
 */
export async function requestSettingChange(
  projectId: string,
  changeType: 'add' | 'modify' | 'delete' | 'merge',
  loreData?: Record<string, unknown>,
  targetLoreId?: string,
  userIntent?: string
): Promise<SettingChangeResult> {
  return await api.post(`${API_BASE}/change`, {
    project_id: projectId,
    change_type: changeType,
    lore_data: loreData,
    target_lore_id: targetLoreId,
    user_intent: userIntent,
  })
}

/**
 * 协商解决冲突
 */
export async function negotiateConflict(
  projectId: string,
  conflictId: string,
  userResponse: string
): Promise<NegotiateResult> {
  return await api.post(`${API_BASE}/negotiate`, {
    project_id: projectId,
    conflict_id: conflictId,
    user_response: userResponse,
  })
}

/**
 * 执行设定变更
 */
export async function executeSettingChange(
  projectId: string,
  requestId: string,
  overrideConflicts: boolean = false
): Promise<ExecuteResult> {
  return await api.post(`${API_BASE}/execute`, {
    project_id: projectId,
    request_id: requestId,
    override_conflicts: overrideConflicts,
  })
}

/**
 * 获取设定摘要
 */
export async function getLoreSummary(projectId: string): Promise<LoreSummary> {
  return await api.get(`${API_BASE}/${projectId}/summary`)
}

/**
 * 获取对话历史
 */
export async function getChatHistory(
  projectId: string,
  sessionId?: string,
): Promise<ConversationHistory> {
  return await api.get(`${API_BASE}/${projectId}/history`, {
    params: { session_id: sessionId },
  })
}

/**
 * 获取待处理冲突
 */
export async function getPendingConflicts(projectId: string): Promise<PendingConflicts> {
  return await api.get(`${API_BASE}/${projectId}/conflicts`)
}

/**
 * 创建或获取会话
 */
export async function createOrGetSession(
  projectId: string,
  mode: 'bootstrap' | 'management' | 'conflict_resolution' = 'management',
  sessionId?: string,
): Promise<{
  success: boolean
  session_id: string
  mode: string
  pending_lores?: PendingLore[]
  pending_characters?: PendingCharacter[]
  pending_hooks?: PendingHook[]
}> {
  return await api.post(`${API_BASE}/${projectId}/session`, null, {
    params: { mode, session_id: sessionId },
  })
}

/**
 * 兼容少数旧调用：读取会话必须仍然使用 POST，避免 GET /session 405。
 */
export async function getOrCreateSession(
  projectId: string,
  mode: 'bootstrap' | 'management' | 'conflict_resolution' = 'management',
  sessionId?: string,
) {
  return createOrGetSession(projectId, mode, sessionId)
}

/**
 * 保存用户确认的设定
 */
export async function resolveCharacterReferences(
  projectId: string,
  references: unknown[],
  provenance?: Record<string, unknown>,
): Promise<CharacterReferenceResolution> {
  return await api.post(`${API_BASE}/resolve-character-references`, {
    project_id: projectId,
    references,
    provenance,
  })
}

export async function savePendingLores(
  projectId: string,
  lores: PendingLore[],
  options?: { sessionId?: string; requestId?: string },
): Promise<{ success: boolean; saved_count: number; message: string; character_reference_resolution?: CharacterReferenceResolution }> {
  return await api.post(`${API_BASE}/save-lores`, {
    project_id: projectId,
    lores,
    session_id: options?.sessionId,
    request_id: options?.requestId,
  })
}

/**
 * 保存用户确认的角色
 */
export async function savePendingCharacters(
  projectId: string,
  characters: PendingCharacter[],
  options?: { sessionId?: string; requestId?: string },
): Promise<{ success: boolean; saved_count: number; message: string }> {
  return await api.post(`${API_BASE}/save-characters`, {
    project_id: projectId,
    characters,
    session_id: options?.sessionId,
    request_id: options?.requestId,
  })
}

/**
 * 保存用户确认的伏笔
 */
export async function savePendingHooks(
  projectId: string,
  hooks: PendingHook[],
  options?: { sessionId?: string; requestId?: string },
): Promise<{ success: boolean; saved_count: number; message: string }> {
  return await api.post(`${API_BASE}/save-hooks`, {
    project_id: projectId,
    hooks,
    session_id: options?.sessionId,
    request_id: options?.requestId,
  })
}

/**
 * 执行设定修改
 */
export async function executeLoreModification(
  projectId: string,
  modification: ImprovementSuggestion
): Promise<{ success: boolean; message?: string; error?: string; character_reference_resolution?: CharacterReferenceResolution }> {
  return await api.post(`${API_BASE}/execute-modification`, {
    project_id: projectId,
    modification,
  })
}

/**
 * 世界观描述结构化分析 (strict)
 */
export interface WorldDescriptionAnalysis {
  power_system: string
  technology_level: string
  history: string
  geography: string
}

export async function analyzeWorldDescription(
  projectId: string,
  description: string
): Promise<{ structured_data: WorldDescriptionAnalysis }> {
  return await api.post(`${API_BASE}/analyze-world`, {
    project_id: projectId,
    description,
  })
}

File: app/api/routes/setting_agent.py
"""
Setting Agent API 路由
设定管理、冲突检测、协商解决
"""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Any, Dict, List, Optional

from app.models.setting_agent import (
    SettingAgentMode,
    SettingChangeType,
    SettingConflict,
)
from app.services.setting_agent_service import get_setting_agent_service

router = APIRouter()


# ==================== 请求模型 ====================

class ChatRequest(BaseModel):
    """聊天请求"""
    project_id: str
    message: str
    context: Optional[Dict[str, Any]] = None


class SettingChangeRequest(BaseModel):
    """设定变更请求"""
    project_id: str
    change_type: str  # add, modify, delete, merge
    lore_data: Optional[Dict[str, Any]] = None
    target_lore_id: Optional[str] = None
    user_intent: Optional[str] = None


class NegotiateRequest(BaseModel):
    """协商请求"""
    project_id: str
    conflict_id: str
    user_response: str


class ExecuteChangeRequest(BaseModel):
    """执行变更请求"""
    project_id: str
    request_id: str
    override_conflicts: bool = False


class SavePendingLoresRequest(BaseModel):
    """保存待确认设定请求"""
    project_id: str
    lores: List[Dict[str, Any]]


class SavePendingCharactersRequest(BaseModel):
    """保存待确认角色请求"""
    project_id: str
    characters: List[Dict[str, Any]]


class ExecuteLoreModificationRequest(BaseModel):
    """执行设定修改请求"""
    project_id: str
    modification: Dict[str, Any]


# ==================== API 端点 ====================

@router.post("/chat")
async def chat_with_setting_agent(request: ChatRequest):
    """
    与 Setting Agent 聊天

    可以用于：
    - 询问设定相关的问题
    - 请求添加/修改设定
    - 获取设定建议

    注意：返回的 pending_lores 需要用户确认后才保存到数据库
    """
    service = get_setting_agent_service()

    try:
        result = await service.chat(
            project_id=request.project_id,
            message=request.message,
            context=request.context,
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/save-lores")
async def save_pending_lores(request: SavePendingLoresRequest):
    """
    保存用户确认的设定

    在设定助手的确认弹窗中，用户确认后调用此接口保存设定
    """
    service = get_setting_agent_service()

    try:
        saved_count = await service.save_pending_lores(
            project_id=request.project_id,
            lores=request.lores,
        )
        return {
            "success": True,
            "saved_count": saved_count,
            "message": f"成功保存 {saved_count} 个设定" if saved_count > 0 else "没有需要保存的设定"
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/save-characters")
async def save_pending_characters(request: SavePendingCharactersRequest):
    """
    保存用户确认的角色

    在设定助手的确认弹窗中，用户确认后调用此接口保存角色
    """
    service = get_setting_agent_service()

    try:
        saved_count = await service.save_pending_characters(
            project_id=request.project_id,
            characters=request.characters,
        )
        return {
            "success": True,
            "saved_count": saved_count,
            "message": f"成功保存 {saved_count} 个角色" if saved_count > 0 else "没有需要保存的角色"
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/change")
async def request_setting_change(request: SettingChangeRequest):
    """
    请求设定变更

    会自动进行冲突检测，返回检测结果
    """
    service = get_setting_agent_service()

    try:
        # 转换变更类型
        change_type = SettingChangeType(request.change_type)

        result = await service.process_setting_change(
            project_id=request.project_id,
            change_type=change_type,
            lore_data=request.lore_data,
            target_lore_id=request.target_lore_id,
            user_intent=request.user_intent,
        )

        return {
            "success": True,
            "request": result.model_dump(),
            "has_conflicts": result.has_conflicts,
            "conflicts": [c.model_dump() for c in result.conflicts],
        }
    except ValueError as e:
        raise HTTPException(status_code=400, detail=f"Invalid change_type: {e}")
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/negotiate")
async def negotiate_conflict(request: NegotiateRequest):
    """
    协商解决冲突

    通过多轮对话与用户协商如何解决设定冲突
    """
    service = get_setting_agent_service()

    try:
        result = await service.negotiate(
            project_id=request.project_id,
            conflict_id=request.conflict_id,
            user_response=request.user_response,
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/execute")
async def execute_setting_change(request: ExecuteChangeRequest):
    """
    执行设定变更

    在冲突解决后（或强制覆盖）执行实际的变更
    """
    service = get_setting_agent_service()

    try:
        result = await service.execute_change(
            project_id=request.project_id,
            request_id=request.request_id,
            override_conflicts=request.override_conflicts,
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{project_id}/summary")
async def get_lore_summary(project_id: str):
    """
    获取项目设定摘要

    包括统计信息、宪法级规则、关键实体等
    """
    service = get_setting_agent_service()

    try:
        summary = await service.get_project_lore_summary(project_id)
        return summary.model_dump()
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{project_id}/history")
async def get_conversation_history(project_id: str, limit: int = 50):
    """
    获取 Setting Agent 对话历史
    """
    service = get_setting_agent_service()

    try:
        session = await service.get_or_create_session(project_id)
        history = session.conversation_history[-limit:]
        return {
            "project_id": project_id,
            "session_id": session.id,
            "history": history,
            "total": len(session.conversation_history),
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{project_id}/conflicts")
async def get_pending_conflicts(project_id: str):
    """
    获取待处理的设定冲突
    """
    service = get_setting_agent_service()

    try:
        session = await service.get_or_create_session(project_id)
        return {
            "project_id": project_id,
            "pending_conflicts": [c.model_dump() for c in session.pending_conflicts],
            "count": len(session.pending_conflicts),
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/{project_id}/session")
async def create_or_get_session(project_id: str, mode: str = "management"):
    """
    创建或获取 Setting Agent 会话
    """
    service = get_setting_agent_service()

    try:
        agent_mode = SettingAgentMode(mode)
        session = await service.get_or_create_session(project_id, agent_mode)
        return {
            "success": True,
            "session": session.model_dump(),
        }
    except ValueError as e:
        raise HTTPException(status_code=400, detail=f"Invalid mode: {e}")
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/execute-modification")
async def execute_lore_modification(request: ExecuteLoreModificationRequest):
    """
    执行设定修改

    根据改进建议执行实际的修改操作：
    - optimize: 优化设定内容
    - priority: 调整优先级
    - missing: 添加缺失设定
    - relation: 更新关联关系
    """
    service = get_setting_agent_service()

    try:
        result = await service.execute_lore_modification(
            project_id=request.project_id,
            modification=request.modification,
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
